client/src/main/java/io/pravega/client/segment/impl/SegmentOutputStreamImpl.java

```diff
@@ -365,7 +379,18 @@ private void ackUpTo(long ackLevel) {
             }
             state.releaseIfEmptyInflight();
         }
-        
+
+        private void checkAckLevels(long ackLevel, long previousAckLevel) {
+            checkState(previousAckLevel < ackLevel, "Bad ack from server - previousAckLevel = %s, ackLevel = %s",
```

Comment:
Ideally, this is should be `previousAckLevel + 1 == ackLevel`, but I understand that we are doing it this way because of conditional updates that can fail. I'm wondering though if we can change the semantics of the ack server side slightly. Here is a suggestion.

We `previousEventNumber` should be the latest acknowledged, positively or negatively. As of now, we only have positive acknowledgements. If we make this change to the semantics of acts, then we should be able to strengthen this check.

Reply:
On the server side, the last successfully written event number is stored per segment per writer id. This does not seem possible with the current design. @fpj @tkaitchuck

Reply:
Moizsj is correct, there is nothing recorded on a failed write attempt. So if you hit a different host, say due to a failover, there would be no way of knowing the failed conditional write ever existed from the server's point of view.